### PR TITLE
Yqxu dev

### PR DESCRIPTION
**Pull request by yqxu**
**What this PR does**:

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

**Comment by zhangshen023** [comment]
error "import cycle not allowed" has occurred!

File: go.mod [hunk 1]
```diff
@@ -5,10 +5,17 @@ go 1.14
 require (
 	github.com/apache/dubbo-go v1.5.1
 	github.com/ghodss/yaml v1.0.1-0.20190212211648-25d852aebe32
+	github.com/gin-gonic/gin v1.6.3
```
Comment (zouyx) [inline go.mod:8]: Why import gin?

File: pkg/client/client.go [hunk 2]
```diff
@@ -17,9 +17,61 @@
...
+)
+
 type Client interface {
 	Init() error
 	Close() error
-
 	Call(req *Request) (resp Response, err error)
 }
+
+type ClientPool struct {
```
Comment (zouyx) [inline pkg/client/client.go:35]: `ClientPool` is private or public scope?

File: pkg/context/context.go [hunk 3]
```diff
@@ -25,7 +25,7 @@ import (
 type Context interface {
 	Next()
 	Abort()
-
+	AbortAndLog(string, error)
```
Comment (zouyx) [inline pkg/context/context.go:28]: I don't think add a method for this interface is a good idea.

**Review by zouyx — COMMENTED** [review]
Make the dog silent and fix conflict

File: pkg/client/client.go (hunk 2 above)
Comment (cityiron) [reply pkg/client/client.go:35]: I think `ClientPool` should in private scope,  when use it, just run Call.

File: pkg/client/client.go [hunk 4]
```diff
@@ -17,9 +17,62 @@
 
 package client
 
+import (
+	"errors"
+	"github.com/dubbogo/dubbo-go-proxy/pkg/client/dubbo"
+	"github.com/dubbogo/dubbo-go-proxy/pkg/client/httpclient"
+	"github.com/dubbogo/dubbo-go-proxy/pkg/model"
+	"sync"
```
Comment (zouyx) [inline pkg/client/client.go:25]: split pkg

File: pkg/client/httpclient/http_response.go [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+	"fmt"
+	"reflect"
+	"strings"
+)
+
+import (
+	"github.com/dubbogo/dubbo-go-proxy/pkg/client"
+)
+
+// NewDubboResponse create dubbo response
```
Comment (zouyx) [inline pkg/client/httpclient/http_response.go:13]:
```suggestion
// NewDubboResponse creates dubbo response
```

File: pkg/context/base_context.go [hunk 6]
```diff
@@ -45,10 +46,18 @@ func (c *BaseContext) Next() {
 	}
 }
 
+// AbortWithError  filter chain break , filter after the current filter will not executed.
```
Comment (zouyx) [inline pkg/context/base_context.go:49]: Comment is wrong - -

File: pkg/context/context.go [hunk 7]
```diff
@@ -25,7 +25,7 @@ import (
 type Context interface {
 	Next()
 	Abort()
-
+	AbortWithError(string, error)
```
Comment (zouyx) [inline pkg/context/context.go:28]: Add a comment for this method pls

File: pkg/context/context.go (hunk 3 above)
Comment (yqxu) [reply pkg/context/context.go]: renamed

File: pkg/client/request.go [hunk 8]
```diff
@@ -19,19 +19,19 @@ package client
 
 import (
 	"github.com/dubbogo/dubbo-go-proxy/pkg/router"
+	"net/http"
```
Comment (AlexStocks) [inline pkg/client/request.go:22]: move it to the 1st import block.

File: pkg/pool/pool.go [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+
+package pool
+
+import (
+	"errors"
+	"github.com/dubbogo/dubbo-go-proxy/pkg/client"
+	"github.com/dubbogo/dubbo-go-proxy/pkg/client/dubbo"
+	"github.com/dubbogo/dubbo-go-proxy/pkg/client/httpclient"
+	"github.com/dubbogo/dubbo-go-proxy/pkg/config"
+	"sync"
```
Comment (zouyx) [inline pkg/pool/pool.go:26]: split pkg

File: pkg/service/api/discovery_service.go [hunk 10]
```diff
@@ -18,12 +18,11 @@
...
 
 import (
+	"errors"
 	"fmt"
 )
 
-import (
-	"github.com/pkg/errors"
-)
+import ()
```
Comment (zouyx) [inline pkg/service/api/discovery_service.go:25]: remove it？

File: pkg/common/constant/key.go [hunk 11]
```diff
@@ -22,7 +22,7 @@ const (
 	HttpRouterFilter         = "dgp.filters.http.router"
 	HttpApiFilter            = "dgp.filters.http.api"
 	HttpDomainFilter         = "dgp.filters.http.domain"
-	HttpTransferDubboFilter  = "dgp.filters.http_transfer_dubbo"
+	RemoteCallFilter         = "dgp.filters.remote_call"
```
Comment (cityiron) [inline pkg/common/constant/key.go:25]: I think name remote is enough

File: pkg/client/request.go [hunk 12]
```diff
@@ -17,21 +17,24 @@
...
-	Body   []byte
-	Header map[string]string
-	API    *router.API
+	IngressRequest *http.Request
+	API            *router.API
 }
 
 // NewRequest create a request
-func NewRequest(b []byte, api *router.API) *Request {
+func NewRequest(request *http.Request, api *router.API) *Request {
```
Comment (cityiron) [inline pkg/client/request.go:35]: I think request is too long, req is better

File: pkg/client/dubbo/dubbo_client.go [hunk 13]
```diff
@@ -92,9 +93,17 @@ func (dc *DubboClient) Close() error {
 // Call invoke service
 func (dc *DubboClient) Call(r *client.Request) (resp client.Response, err error) {
 	dm := r.API.Method.IntegrationRequest
-	gs := dc.Get(dm.Interface, dm.Version, dm.Group, dm)
+	gs := dc.Get(dm.Interface, dm.DubboBackendConfig.Version, dm.Group, dm)
 
 	var reqData []interface{}
+	var indate []byte
```
Comment (williamfeng323) [inline pkg/client/dubbo/dubbo_client.go:99]: inData？

File: pkg/context/base_context.go [hunk 14]
```diff
@@ -45,10 +46,18 @@ func (c *BaseContext) Next() {
...
 
+// Abort  filter chain break , filter after the current filter will not executed.
 func (c *BaseContext) Abort() {
 	c.Index = abortIndex
 }
 
+// AbortWithError  filter chain break , filter after the current filter will not executed. And log will print.
+func (c *BaseContext) AbortWithError(message string, err error) {
+	c.Index = abortIndex
+	log.Println(message, err) //TODO print stack
```
Comment (williamfeng323) [inline pkg/context/base_context.go:57]: use the logger in pkg/logger?

File: pkg/filter/remote_call_filter.go [hunk 15]
```diff
@@ -31,40 +27,40 @@ import (
...
 	return func(c context.Context) {
-		doDubbo(c.(*http.HttpContext))
+		doRemoteCall(c.(*http.HttpContext))
 	}
 }
 
-func doDubbo(c *http.HttpContext) {
+func doRemoteCall(c *http.HttpContext) {
 	api := c.GetAPI()
+	cl, e := pool.SingletonPool().GetClient(api.Method.IntegrationRequest.RequestType)
```
Comment (williamfeng323) [inline pkg/filter/remote_call_filter.go:51]: defer pool.put to release the client?

Comment (yqxu) [reply pkg/filter/remote_call_filter.go:51]: This client is singleton , May using same instance in different G will cause race condition?

File: pkg/client/httpclient/http_client.go [hunk 16]
```diff
@@ -0,0 +1,156 @@
...
+		})
+	}
+	return _httpClient
+}
+
+// NewHTTPClient create dubbo client
+func NewHTTPClient() *HTTPClient {
+	return &HTTPClient{
+		mLock:              sync.RWMutex{},
+		GenericServicePool: make(map[string]*dg.GenericService),
```
Comment (AlexStocks) [inline pkg/client/httpclient/http_client.go:86]: pls set the map init length.

File: pkg/client/httpclient/http_client.go [hunk 17]
```diff
@@ -0,0 +1,156 @@
...
+
+	var urlStr = r.API.IntegrationRequest.HTTPBackendConfig.Protocol + "://" + r.API.IntegrationRequest.HTTPBackendConfig.TargetURL
+	var httpClient = &http.Client{Timeout: 5 * time.Second}
+	var request = r.IngressRequest.Clone(context.Background())
+	request.URL, _ = url.ParseRequestURI(urlStr)
+	//TODO header replace, url rewrite....
+
+	var tmpRet, e = httpClient.Do(request)
+	var ret = client.Response{Data: tmpRet}
+	return ret, e
```
Comment (AlexStocks) [inline pkg/client/httpclient/http_client.go:115]: so many vars！！！

**Review by williamfeng323 — APPROVED** [review]

**Review by AlexStocks — APPROVED** [review]

**Review by zhangshen023 — APPROVED** [review]

**Pull request merged**